### PR TITLE
Feature/add maml meta test

### PR DESCRIPTION
**Pull request by Clyde21c**
## PR Description

This PR is baseline for further experiment of MAML algorithm implementation.
- determine training structures to follow the learn2learn repo
- add meta-update
- add Linear-feature baseline
- compute advantage with newly fitted value
- change training configs to follow the official repo of MAML
- add multi-task buffer

## Related Issues

#31 #44 #47 

## Checklist

- [x] **Assignees** ran the `make format`.
- [x] **Assignees** ran the `make lint`.
- [x] **Assignees** checked the runnability of the code.
- [ ] **Reviewers** checked the runnability of the code.
- [ ] **Reviewers** checked plausibility of the training structure

## Experimental Results

TBU

## TODO

1. ~~Add Meta-training~~
2. ~~Add Meta-test~~
3. Experiment MAML with additional components for stabilization such as KL divergence and tanh output
4. Experiment MAML with different HP
6. Match code style and added functions with `RL^2` and `PEARL`

### CODE REVIEW

**Comment by Clyde21c** [comment]
@dongminlee94 동민님. 본격적으로 코드 구조를 바꾸는 실험을 해보기에 앞서 논의가 필요할것 같아 PR 날립니다.
어디까지나 논의를 위한 merge이며, 아직 온전히 학습이 되지 않는다는 점 고려부탁드립니다.

- 우선, l2l, rllib 모두 trpo와 ppo를 위해 outer-loop에서 사용할 데이터를 저장해두고 여러번 학습을 합니다. 이를 위해 버퍼 구조가 변경이 필요해 보입니다. 
- 또한 MAML의 경우 gradient의 폭이 커서 KL loss의 사용이 필요한지의 논의가 필요합니다. 이 경우 마찬가지로 버퍼 코드가 많이 달라질것 같습니다. 
- rllib은 MAML에서 adapted_trajectory를 사용하지 않도록 구현을 한것으로 보입니다. 이부분은 원 논문의 수도코드와 방향이 크게 달라 채택을 하기 어려울것 같습니다.
- l2l과 같이 baseline을  fit하는 구조로 갈 경우, 마찬가지로 버퍼 및 샘플러 코드의 변경이 필요해 보입니다.

**Comment by Clyde21c** [comment]
@dongminlee94 policy의 구조를 변경하였습니다. 예전에 log_std의 initial value가 중요한 코드였다면, RL2에서 사용한 std head를 추가하는 방법으로 통일시켰습니다.

**Comment by dongminlee94** [comment]
> @dongminlee94 동민님. 본격적으로 코드 구조를 바꾸는 실험을 해보기에 앞서 논의가 필요할것 같아 PR 날립니다.
> 마지막 HP튜닝을 앞둔 코드가 아니라 어디까지나 학습 구조 변경 논의를 위한 merge이며, 때문에 아직 온전히 학습이 되지 않는다는 점 고려부탁드립니다.
> 
> * 우선, l2l, rllib 모두 trpo와 ppo를 위해 outer-loop에서 사용할 데이터를 저장해두고 여러번 재활용하여 학습을 합니다. 이를 위해 버퍼 구조가 변경이 필요해 보입니다.
> * 또한 MAML의 경우 gradient의 폭이 커서 KL loss의 사용이 필요한지의 논의가 필요합니다. 이 경우 마찬가지로 RL2와 버퍼 코드가 많이 달라질것 같습니다.
> * rllib은 MAML에서 adapted_trajectory를 사용하지 않도록 구현을 한것으로 보입니다. 이부분은 원 논문의 수도코드와 방향이 크게 달라 채택을 하기 어려울것 같습니다.
> * l2l과 같이 baseline을  fit하는 구조로 갈 경우, 마찬가지로 버퍼 및 샘플러 코드의 변경이 필요해 보입니다.

@Clyde21c 네 승현님. 혹시 괜찮으시다면, 9월 7일 화요일 전체 회의 마치고 PR 소개 미팅을 잠깐 가능하신가요? 1차적으로 서로 대화를 통해 PR에 대한 내용을 듣고 코드를 보는것이 더 좋을 것 같습니다.

**Comment by Clyde21c** [comment]
@dongminlee94 네 동민님. 그러시죠.

**Comment by Clyde21c** [comment]
@dongminlee94 화요일 논의한 내용 반영하였습니다. 학습은 진행중이고, 검토 부탁드리겠습니다. 수정 커밋이 많아 헷갈리는 부분이 있을 수 있으니 말씀주시면 추가 설명드리겠습니다.

File: src/maml/algorithm/buffer.py [hunk 1]
```diff
@@ -26,77 +66,85 @@ def __init__(  # pylint: disable=too-many-arguments
...
-            )
-            running_advant = (
-                running_tderror + self.gamma * self.lamda * (1 - self._dones[t]) * running_advant
-            )
-            self._advants[t] = running_advant
-            prev_value = self._values[t]
+            running_tderror = rewards[t] + gamma * (1 - dones[t]) * prev_value - values[t]
+            running_advant = running_tderror + gamma * lamda * (1 - dones[t]) * running_advant
+            advants[t] = running_advant
+            prev_value = values[t]
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/buffer.py:127]: RL2와 통일성을 위해 `buffer.py에 있는 `compute_gae`를 불러다 쓰기 위함입니다. 합의만 된다면 gae 계산부분을 ppo.py로 옮겨올까 합니다. 어떻게 생각하시나요?

File: src/maml/algorithm/meta_learner.py [hunk 2]
```diff
@@ -55,27 +55,27 @@ def __init__(  # pylint: disable=too-many-arguments
...
+            num_episodes=(self.num_adapt_epochs + 1),  # [num of adapatation for train] + [validation]
+            max_size=self.num_samples,
             device=device,
         )
 
-        self.meta_optimizer = torch.optim.Adam(
-            list(self.agent.policy.parameters()),
-            lr=config["learning_rate"],
+        self.test_buffer = Buffer(
+            observ_dim=observ_dim, action_dim=action_dim, max_size=self.num_samples, device=device
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/meta_learner.py:68]: Train 중에는 PPO 를 위해 모든 task에 대한 결과를 저장해두고 다시 사용하는 반면, test 중에는 task adaptation 후에 데이터를 다시 쓰지 않기 때문입니다.

File: src/maml/algorithm/meta_learner.py [hunk 3]
```diff
@@ -88,64 +88,224 @@ def __init__(  # pylint: disable=too-many-arguments
...
+        for iteration in range(self.num_iterations):
+            start_time = time.time()
 
-            print("=============== Iteration {} ===============".format(iteration))
+            print(f"=============== Iteration {iteration} ===============")
             # Sample tasks randomly from train tasks distribution.
-            for _ in range(self.num_sample_tasks):
-                index = np.random.randint(len(self.train_tasks))
+            indices = np.random.randint(len(self.train_tasks), size=self.num_sample_tasks)
+            for cur_task, index in enumerate(indices):
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/meta_learner.py:149]: 네 buffer에 저장하기 위한 index와 env_reset을 하기위한 index가 다르기 때문에 두 index를 분리해주기위하여 enumerate를 썼습니다

File: src/maml/algorithm/networks.py [hunk 4]
```diff
@@ -46,29 +49,61 @@ def forward(self, x):
...
+        reg = self.reg * torch.eye(features.size(1))
+        A = features.t() @ features + reg
+        b = features.t() @ returns
+        coeffs, _ = torch.lstsq(b, A)
+        self.linear.weight.data = coeffs.data.t()
+
+    def forward(self, obs):
+        """Get value when observations are given"""
+        features = LinearValue.features(obs)
+        return self.linear(features)
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/networks.py:85]: pylint에서 self.features를 쓰려고 하니 feature 내부에 self가 쓰이지 않는다는 오류가 떠서 어쩔 수 없이 decorator를 썼습니다. 다른 방법이 있을까요?

File: src/maml/algorithm/networks.py [hunk 5]
```diff
@@ -46,29 +49,61 @@ def forward(self, x):
...
+
+
+class TanhGaussianPolicy(MLP):
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/networks.py:88]: 이부분은 좀 고민인게,  tanh를 쓰지 않으니 학습이 엄청 불안정하게 나오긴합니다. 지금은 maml PPO 구조를 어떻게 짤지 부터 확실하지 않은 상황이긴하나,  나중에 필요하면 추가해야할수도 있을것 같습니다

File: src/maml/algorithm/ppo.py [hunk 6]
```diff
@@ -16,58 +15,70 @@ def __init__(  # pylint: disable=too-many-arguments
...
-        value_batch = self.vf(obs_batch)
-        value_loss = F.mse_loss(value_batch, return_batch)
-
-        self.vf_optimizer.zero_grad()
-        value_loss.backward()
-        self.vf_optimizer.step()
+        # Compute advantage with standard linear feature baseline (Daun el al. 2016)
+        self.baseline.fit(obs_batch, return_batch)
+        value_batch = self.baseline(obs_batch).detach()
+        advant_batch = Buffer.compute_gae(reward_batch, done_batch, value_batch).to(self.device)
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/ppo.py:65]: 위에 관련 리뷰와 마찬가지로 RL2와의 통일성을 생각하지 않고 `compute gae`를 PPO로 넣으면 어떨까요?

File: src/maml/algorithm/ppo.py [hunk 7]
```diff
@@ -16,58 +15,70 @@ def __init__(  # pylint: disable=too-many-arguments
...
+        else:
+            ratio = torch.exp(new_log_prob_batch - log_prob_batch)
+            surr_loss = ratio * advant_batch
+            if clip_loss:
+                clipped_loss = (
+                    torch.clamp(ratio, 1 - self.clip_param, 1 + self.clip_param) * advant_batch
+                )
+                policy_loss = -torch.min(surr_loss, clipped_loss).mean()
+            else:
+                policy_loss = -surr_loss.mean()
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/algorithm/ppo.py:81]: ray에서 inner_adaptation 중에는 clip_loss를 안쓰는 반면, meta-update에서는 clip_loss를 쓰도록 되어있어, 해당 부분이 구현된 부분입니다.

File: src/maml/configs/vel_target_config.yaml [hunk 8]
```diff
@@ -3,40 +3,34 @@
...
-hidden_dims: [300, 300, 300]
+hidden_dim: 256
 
 maml_params:
     # Number of training iterations
-    train_iters: 200
+    num_iterations: 1000
     # Number of task samples for training
-    num_sample_tasks: 5
+    num_sample_tasks: 40
```
[the remark this answers is not in the document]
Comment (Clyde21c) [reply src/maml/configs/vel_target_config.yaml:18]: MAML 공식 레포에서 사용한 값을 기준으로 잡고자 같은 값을 사용했습니다

**Comment by Clyde21c** [comment]
@dongminlee94 동민님, MAML에 PPO를 적용하려면 TRPO를 쓰는 learn2learn과는 meta-update의 구조가 많이 달라져야 할것 같아요. Ray와 learn2learn 중간쯤 방법으로 가야할것 같아 시도를 해봤는데, 성능이 잘 안나오네요. 구현보단 알고리즘 적인 문제라 혼자 고민하기 벅찬 부분같아 시간 되시면 한번 더 논의를 하면 좋을 것 같습니다.

**Comment by Clyde21c** [comment]
This PR is closed as the branch is deprecated.
